validator/client/src/main/java/tech/pegasys/teku/validator/client/ActiveKeyManager.java

```diff
@@ -214,57 +271,141 @@ DeleteKeyResult deleteExternalValidator(final Validator activeValidator) {
   public List<PostKeyResult> importValidators(
       final List<String> keystores,
       final List<String> passwords,
-      final Optional<SlashingProtectionImporter> slashingProtectionImporter) {
-    final List<PostKeyResult> importResults = new ArrayList<>();
+      final Optional<SlashingProtectionImporter> slashingProtectionImporter,
+      final Optional<DoppelgangerDetector> maybeDoppelgangerDetector,
+      final Optional<DoppelgangerDetectionAction> maybeDoppelgangerDetectionAction) {
+    final List<Pair<Optional<BLSPublicKey>, PostKeyResult>> importResults = new ArrayList<>();
     boolean reloadRequired = false;
     for (int i = 0; i < keystores.size(); i++) {
       importResults.add(
           importValidatorFromKeystore(
               keystores.get(i), passwords.get(i), slashingProtectionImporter));
-      if (importResults.get(i).getImportStatus() == ImportStatus.IMPORTED) {
+      if (importResults.get(i).getValue().getImportStatus() == ImportStatus.IMPORTED) {
         reloadRequired = true;
       }
     }
     if (reloadRequired) {
-      validatorTimingChannel.onValidatorsAdded();
+      if (maybeDoppelgangerDetector.isPresent()) {
+        maybeDoppelgangerDetector
+            .get()
+            .performDoppelgangerDetection(filterImportedPubKeys(importResults))
+            .thenAccept(
```

Comment:
Haven't we already loaded the keys and added them to the `OwnedValidator` set at this point? It seems too late to be enabling doppelgänger.  The call to `validatorTimingChannel.onValidatorsAdded` is just a trigger to immediately reload duties for validators, but duties could be loaded at any point (e.g by a reorg). What stops the validator's still performing doppelgänger detection from having duties calculated?

Reply:
Right. I implemented a new approach: Instead of loading the validators and then deleting the detected doppelgangers, we now load the keys without adding the validators (Hence we're sure they don't get their duties scheduled). Once the DG check is finished we add the validators (and ignore the doppelganger keys)